### PR TITLE
feat: add promotion to order

### PR DESCRIPTION
**Pull request by frobel**
- possibility to add a promotion as LineItem to an order via overrides

### CODE REVIEW

**Comment by frobel** [comment]
WIP

File: src/services/TestDataService.ts [hunk 1]
```diff
@@ -1153,6 +1156,13 @@ export class TestDataService {
         return Object.assign({}, basicCustomer, overrides);
     }
 
+    isProduct(item: Product | Promotion): item is Product {
+        return (item as Product).productNumber !== undefined;
+    }
+    isPromotion(item: Product | Promotion): item is Promotion {
```
Comment (Phil23) [inline src/services/TestDataService.ts:1163]: Please add an empty line between methods.

File: src/services/TestDataService.ts [hunk 2]
```diff
@@ -1174,11 +1184,28 @@ export class TestDataService {
         const shippingDateTime = this.convertDateTime(shippingDate);
 
         let totalPrice = 0;
-        const lineItemProducts: Record<string, unknown>[] = [];
+        const orderLineItems: Record<string, unknown>[] = [];
         lineItems.forEach((lineItem) => {
-            lineItemProducts.push(this.getBasicProductLineItemStruct(lineItem));
-            totalPrice += lineItem.product.price[0].gross * (lineItem.quantity || 1);
-        });
+            if(this.isProduct(lineItem.product)){
```
Comment (Phil23) [inline src/services/TestDataService.ts]: Code Style: Whitespace missing after `if`.

Comment (Phil23) [inline src/services/TestDataService.ts]: Code Style: Whitespace missing before `{`

File: src/services/TestDataService.ts [hunk 3]
```diff
@@ -1174,11 +1184,28 @@ export class TestDataService {
...
+                const product = lineItem.product;
+                orderLineItems.push(this.getBasicProductLineItemStruct(lineItem));
+                totalPrice += product.price[0].gross * (lineItem.quantity || 1);
+            }
+            if(this.isPromotion(lineItem.product)){
```
Comment (Phil23) [inline src/services/TestDataService.ts]: Please check code style.

File: src/services/TestDataService.ts [hunk 4]
```diff
@@ -1306,19 +1333,23 @@ export class TestDataService {
         return Object.assign({}, basicOrder, overrides);
     }
 
-    getBasicProductLineItemStruct(lineItem: SimpleLineItem) {
-        const unitPrice = lineItem.product.price[0].gross || 10;
+    getBasicProductLineItemStruct(lineItem: SimpleLineItem ) {
+        if (!this.isProduct(lineItem.product)){
+            console.error('Error: Object is not of type Product');
+        }
+        const product = lineItem.product as Product;
```
Comment (Phil23) [inline src/services/TestDataService.ts:1343]: Code Style: Please add an empty line after the `if`.

File: src/services/TestDataService.ts [hunk 5]
```diff
@@ -1350,6 +1381,50 @@ export class TestDataService {
         };
     }
 
+    getBasicPromotionLineItemStruct(lineItem: SimpleLineItem) {
+        if (!this.isPromotion(lineItem.product)){
+            console.error('Error: Object is not of type Promotion');
+        }
+        const promotion = lineItem.product as Promotion;
```
Comment (Phil23) [inline src/services/TestDataService.ts:1392]: Code Style: Please add an empty line after the `if`.

File: tests/TestDataService.spec.ts [hunk 6]
```diff
@@ -23,13 +23,16 @@ test('Data Service', async ({
     expect(customer.firstName).toEqual('Luke');
     expect(customer.lastName).toEqual('Skywalker');
 
+    const promotion = await TestDataService.createPromotionWithCode({ code: 'myCode' });
+    expect(promotion.code).toEqual('myCode');
+
     const order = await TestDataService.createOrder(
-        [{ product, quantity: 5 }],
+        [{ product:product, quantity: 5 }, { product:promotion, quantity: 1 }],
```
Comment (Phil23) [inline tests/TestDataService.spec.ts]: Code Style: Add a whitespace after `:`.

**Review by Phil23 — CHANGES_REQUESTED** [review]

File: src/services/ShopwareDataHelpers.ts [hunk 7]
```diff
@@ -254,4 +255,28 @@ export function extractIdFromUrl(url: string): string | null {
...
+            filter: [{
+                type: 'equals',
+                field: 'id',
+                value: promotionId,
+            }],
+        },
+    });
+
+    const { data: promotion } = (await resp.json()) as { data: Promotion[] };
+    return promotion;
```
Comment (Phil23) [inline src/services/ShopwareDataHelpers.ts]: As this method is explicitly searching for a specific promotion, it should not return an array but a single promotion.

**Review by Phil23 — CHANGES_REQUESTED** [review]

File: src/services/ShopwareDataHelpers.ts (hunk 7 above)
Comment (frobel) [reply src/services/ShopwareDataHelpers.ts]: added that the method returns just one promotion.

File: src/services/TestDataService.ts [hunk 8]
```diff
@@ -1174,10 +1185,28 @@ export class TestDataService {
         const shippingDateTime = this.convertDateTime(shippingDate);
 
         let totalPrice = 0;
-        const lineItemProducts: Record<string, unknown>[] = [];
+        const orderLineItems: Record<string, unknown>[] = [];
         lineItems.forEach((lineItem) => {
-            lineItemProducts.push(this.getBasicProductLineItemStruct(lineItem));
-            totalPrice += lineItem.product.price[0].gross * (lineItem.quantity || 1);
+            if(this.isProduct(lineItem.product)) {
```
Comment (Phil23) [inline src/services/TestDataService.ts]: Code Style

File: src/services/TestDataService.ts [hunk 9]
```diff
@@ -1174,10 +1185,28 @@ export class TestDataService {
...
+                const product = lineItem.product;
+                orderLineItems.push(this.getBasicProductLineItemStruct(lineItem));
+                totalPrice += product.price[0].gross * (lineItem.quantity || 1);
+            }
+
+            if(this.isPromotion(lineItem.product)) {
```
Comment (Phil23) [inline src/services/TestDataService.ts]: Code Style

**Review by Phil23 — CHANGES_REQUESTED** [review]

**Review by Phil23 — DISMISSED** [review]

**Review by Phil23 — APPROVED** [review]

**Pull request merged**